internal/mesh/internal/controllers/gatewayproxy/controller.go

```diff
@@ -85,14 +87,119 @@ func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req c
 		return r.reconcileMeshGatewayProxyState(ctx, dataFetcher, workload, rt, req)
 	case apigateways.GatewayKind:
 		rt.Logger.Trace("workload is a api-gateway; reconciling", "workload", workloadID, "workloadData", workload.Data)
-		// TODO: NET-735 -- implement api-gateway reconciliation
-		return nil
+		return r.reconcileAPIGatewayProxyState(ctx, dataFetcher, workload, rt, req)
 	default:
 		rt.Logger.Trace("workload is not a gateway; skipping reconciliation", "workload", workloadID)
 		return nil
 	}
 }
 
+func (r *reconciler) reconcileAPIGatewayProxyState(ctx context.Context, dataFetcher *fetcher.Fetcher, workload *resource.DecodedResource[*pbcatalog.Workload], rt controller.Runtime, req controller.Request) error {
+	proxyStateTemplate, err := dataFetcher.FetchProxyStateTemplate(ctx, req.ID)
+	if err != nil {
+		rt.Logger.Error("error reading proxy state template", "error", err)
+		return nil
+	}
+
+	if proxyStateTemplate == nil {
+		req.ID.Uid = ""
+		rt.Logger.Trace("proxy state template for this gateway doesn't yet exist; generating a new one")
+	}
+
+	gwID := &pbresource.ID{
+		Name:    workload.Data.Identity,
+		Type:    pbmesh.APIGatewayType,
+		Tenancy: workload.Id.Tenancy,
+	}
+
+	apiGateway, err := dataFetcher.FetchAPIGateway(ctx, gwID)
+	if err != nil {
+		rt.Logger.Error("error reading the associated api gateway", "error", err)
+		return err
+	}
+
+	if apiGateway == nil {
+		rt.Logger.Trace("api gateway doesn't exist; skipping reconciliation", "apiGatewayID", gwID)
+		return nil
+	}
+
+	allTCPRoutes, err := dataFetcher.FetchAllTCPRoutes(ctx, req.ID.Tenancy)
+	if err != nil {
+		rt.Logger.Error("error reading the associated tcp routes", "error", err)
+		return err
+	}
+
+	if len(allTCPRoutes) == 0 {
+		rt.Logger.Trace("no tcp routes found for this gateway", "apiGatewayID", gwID)
+	}
+
+	services := make([]*pbcatalog.Service, 0)
+
+	referencedTCPRoutes := make([]*pbmesh.TCPRoute, 0)
+	for _, tcpRoute := range allTCPRoutes {
+		for _, parentRef := range tcpRoute.Data.ParentRefs {
+			if parentRef.Ref.Name == apiGateway.Id.Name {
```

Comment:
#nit Seems like it would be safer to assert the entire reference here instead of just the name

something like
```go
if resource.EqualReference(parentRef.Ref, resource.ReferenceFromReferenceOrID(apiGateway.Id)) {
```